src/main/java/uk/nhs/hee/trainee/details/service/NtnGenerator.java

```diff
@@ -0,0 +1,303 @@
+import uk.nhs.hee.trainee.details.model.Curriculum;
+import uk.nhs.hee.trainee.details.model.PersonalDetails;
+import uk.nhs.hee.trainee.details.model.ProgrammeMembership;
+import uk.nhs.hee.trainee.details.model.TraineeProfile;
+
+/**
+ * A service for handling trainee NTNs/DRNs.
+ */
+@Slf4j
+@Service
+public class NtnGenerator {
+
+  /**
+   * Populate the NTN for all programme memberships in the trainee profile.
+   *
+   * @param traineeProfile The trainee profile to populate with NTNs.
+   */
+  public void populateNtns(TraineeProfile traineeProfile) {
+    PersonalDetails personalDetails = traineeProfile.getPersonalDetails();
+
+    if (isExcluded(personalDetails)) {
+      return;
+    }
+
+    traineeProfile.getProgrammeMemberships().forEach(pm -> populateNtn(personalDetails, pm));
+  }
+
+  /**
+   * Populate the NTN for the given programme membership.
+   *
+   * @param personalDetails     The personal details to use for NTN generation.
+   * @param programmeMembership The programme membership to generate the NTN for.
+   */
+  private void populateNtn(PersonalDetails personalDetails,
+      ProgrammeMembership programmeMembership) {
+    log.info("Populating NTN for programme membership '{}'.", programmeMembership.getTisId());
+
+    if (isExcluded(programmeMembership)) {
+      return;
+    }
+
+    String parentOrganization = getParentOrganization(programmeMembership);
+    String specialtyConcat = getSpecialtyConcat(programmeMembership);
+    String referenceNumber = getReferenceNumber(personalDetails);
+    String suffix = getSuffix(programmeMembership);
+    String ntn = parentOrganization + "/" + specialtyConcat + "/" + referenceNumber + "/" + suffix;
+    programmeMembership.setNtn(ntn);
+    log.info("Populated NTN: {}.", ntn);
+  }
+
+  /**
+   * Get the parent organization for the given programme membership.
+   *
+   * @param programmeMembership The programme membership to calculate the parent organization for.
+   * @return The calculated parent organization.
+   */
+  private String getParentOrganization(ProgrammeMembership programmeMembership) {
+    String managingDeanery = programmeMembership.getManagingDeanery();
+    log.info("Calculating parent organization for managing deanery '{}'.", managingDeanery);
+
+    String parentOrganization = managingDeanery == null ? null : switch (managingDeanery) {
+      case "Defence Postgraduate Medical Deanery" -> "TSD";
+      case "Health Education England East Midlands" -> "EMD";
+      case "Health Education England East of England" -> "EAN";
+      case "Health Education England Kent, Surrey and Sussex" -> "KSS";
+      case "Health Education England North Central and East London",
+          "Health Education England South London",
+          "Health Education England North West London",
+          "London LETBs" -> "LDN";
+      case "Health Education England North East" -> "NTH";
+      case "Health Education England North West" -> "NWE";
```

Comment:
In the SQL I'm seeing a few more conditions here:
```
when pgm.ManagingDeanery = 'Health Education England North West' and pgm.ProgrammeName like '%General Practice%' then 'NWE'
				when pgm.ManagingDeanery = 'Health Education England North West' and pgm.ProgrammeNumber like 'NWD%' then 'NWN'
				when pgm.ManagingDeanery = 'Health Education England North West' and pgm.ProgrammeNumber like 'MER%' then 'MER'
```

Reply:
That's the old SQL, north west use a single code now.

Reply:
Ah, right you are. Ok, working from the one attached to the ticket now ('as of 18-06-2024') 😁